pytorch_lightning/profiler/pytorch.py

```diff
@@ -448,68 +613,242 @@ def on_trace_ready(profiler):
-        init_parameters = inspect.signature(profiler.__init__).parameters
-        kwargs = {k: v for k, v in self._profiler_kwargs.items() if k in init_parameters}
-        return profiler(**kwargs)
+            self._transit_profiler(KinetoProfilerState.START)
 
-    def _cache_functions_events(self) -> None:
-        if self._emit_nvtx:
+    def _start_action(self, action_name: str) -> None:
+        self._action_step_num[action_name] += 1
+
+        if self._lightning_module is None:
+            # we don't call prepare_trace/start_trace if the lightning_module is None
+            # like the following scenarios
+            # with pytorch_profiler.profile("a"):
+            #   a = torch.ones(42)
             return
-        self.function_events = self.profiler.events() if _KINETO_AVAILABLE else self.profiler.function_events
 
-    def _delete_profilers(self) -> None:
+        total_steps = self.get_total_steps(action_name)
+        if total_steps is None or total_steps < self.profile_steps:
+            # if the total_step is None, it means the action_name is not belong to
+            # any train/validation/test/prediction steps
+            warning_cache.warn(
+                "The PyTorch Profiler default schedule will be overridden as there is not enough "
+                "steps to properly record traces."
+            )
+            self._override_steps = True
+            # No need to call prepare_trace & start_trace here
+            # since the profiler is already started when created.
+            return
+
+        step_num = self._action_step_num[action_name]
+        if step_num == self._wait_step:
+            # warm up
+            self._transit_profiler(KinetoProfilerState.WARMUP)
+        elif step_num >= self._wait_step + self._warmup_step and step_num < self.profile_steps:
+            # begin profile
+            self._transit_profiler(KinetoProfilerState.START)
+
+    def _stop_action(self, action_name: str) -> None:
+        if self.profiler is not None and any(action_name.endswith(func) for func in self.STEP_FUNCTIONS):
+            # Save the first action name in order to save the profile data
+            # when destroying profiler like what the original on_trace_ready did.
+            if self._start_action_name is None:
+                self._start_action_name = action_name
+
+            if self._action_step_num[action_name] == self.profile_steps:
+                self._transit_profiler(KinetoProfilerState.NONE)
+                if self.dirpath is not None:
+                    self._save_result(action_name)
+                else:
+                    rank_zero_warn("The PyTorchProfiler failed to export trace as `dirpath` is None")
+
+    def _delete_profiler(self) -> None:
         if self.profiler is not None:
-            self.profiler.__exit__(None, None, None)
+            if not _KINETO_AVAILABLE or self._emit_nvtx:
+                self.profiler.__exit__(None, None, None)
```

Comment:
Could this be?

```suggestion
                self.profiler.stop_trace()
```

Reply:
No, the stop_trace only available in new API. the older has to use __exit__.